### ISSUE TITLE
TaskExtensions is not in the LanguageExt namespace

### ISSUE THREAD

**Issue by mattfurness** [louthy/language-ext#445]
I recently added Langauge Ext to an existing codebase and ran into an issue where it started calling [`Select`](https://github.com/louthy/language-ext/blob/v3.0.5/LanguageExt.Core/DataTypes/Task/Task.Extensions.cs#L55) from `TaskExtensions` instead of an extension method in a namespace that I had explicitly imported because the `TaskExtensions` class is [not in any namespace](https://github.com/louthy/language-ext/blob/v3.0.5/LanguageExt.Core/DataTypes/Task/Task.Extensions.cs#L11).  It was a subtle issue that took a little while to track down because I was not explicitly importing anything from Language Ext.

I was just wondering if not putting `TaskExtensions` in the `LangaugeExt` namespace was intentional, or was just accidentally missed. 

Thanks for all your effort on this library @louthy.

**Comment by louthy** [comment]
@mattfurness  Hi Matt, it was a conscious decision to do this, because:

* Really there's only one way to compose Tasks, which is the way that the `async` keyword does it: continue on success, bail out on fail/cancel.  So replicating that in `Select` and `SelectMany` made sense.
* It also made sense that it should be a feature of `Task` already and not something to explicitly opt in to.  i.e. `Task` should be a `Task` monad, it just made sense.

I can see the issue though, mostly when I've done this in the past (putting the extensions in the global namespace) it has been to extend types that belong to lang-ext, so I can totally see your point that it can cause clashes with other implementations.

Obviously moving it to the `LanguageExt` namespace is what needs to happen.  Although also obviously it sounds like that will still preclude you from using the `LanguageExt` namespace along with your other extensions, but at least you'll have the choice.  

Out of interest what are you doing in your version of `Select` for `Task`?  Is it the same mapping functionality?

**Comment by mattfurness** [comment]
Thank you for the super quick reply @louthy 😄 

Regarding your comment on precluding the use of the namespace:
- Fortunately we don't need to import `LanguageExt` in the file in question, so it wouldn't be an issue in this specific case.
- If I had needed to import the `LanguageExt` namespace and the tests started failing in the way they did it would have been much more obvious. It was the implicitness that caught me out.

Regarding the version of `Select` in the existing project, I'll try to describe it succinctly:
- The function / method signature was similar to `Task<CurProjResult<S>> Select<T, S>(this Task<CurProjResult<T>> task, Func<T, S> func)`. The implementation checks that the wrapped `CurProjResult` is `Succeeded` before calling `func`
- I was a little surprised that the existing function above wasn't the function called considering it appears to be "more specific" but the `TaskExtensions` function was chosen instead.

As you mentioned in your previous comment, I think moving it into the `LanguageExt` namespace is the right move, even though it is a breaking change, simply out of explicitness.

**Comment by bentefay** [comment]
I can provide a bit more detail (I work with @mattfurness).

We have an extension method:

``` C#
namespace Company { 
    public static class Extensions {
        public static Task<Company.Result<S>> Select<T, S>(this Task<Company.Result<T>> task, Func<T, S> func);
    }
}
```

When it comes to extension method overload resolution, if the target method is ambiguous (as it is in this case between Language Ext's `Select` method and ours), the method in the shallower namespace wins. Because the global namespace is shallower than our `Company` namespace, the Language Exts' `Select` method wins. If we put *our* extension method into the global namespace, the compiler can no longer disambiguate the methods and raises an error: `The call is ambiguous between the following methods`.

Long story short, by putting `public static async Task<U> Select<T, U>(this Task<T> self, Func<T, U> map)` in the global namespace, it becomes impossible for consumers of Language Ext to write a `Select` extension method on a `Task` of something more specific.

**Comment by louthy** [comment]
Yep, understood.  If you fancy doing a pull request that I can just deploy when I get back in front of my machine (may be Sunday or Monday) that’d be great.

Paul

**Comment by louthy** [comment]
This is now deployed.  Thanks for the pull request 👍

**Comment by mattfurness** [comment]
No worries, thanks for responding so quickly 😄
